### PR TITLE
Sync with `ab-production-runtime` and add support for podman

### PR DESCRIPTION
**Pull request by nh758**
- #353 Adds Podman Support to `dockerImageUpdate.js`

- Also merges in the `ab-production-runtime` repo with the plan to **depreciate** that, since I don't want to have to maintain both.
    - This runtime should work for both development and production. It can be installed using `ab-cli` (requires https://github.com/digi-serve/ab-cli/pull/67), or cloned with setup steps similar to what we did in `ab-production-runtime` (see updated `README`).
## Release Notes

- combine `ab-production-runtime` so we don't need to maintain multiple runtimes
- improve podman compatibility, including no downtime image updates

### CODE REVIEW

**Review by joshAppDev — APPROVED** [review]
Looks good.

I did not know about the `podman compose` command. Since it delegates to either `podman-compose` or `docker-compose` depending on what's installed on the system, I wonder if there would be weird interactions for the combination of podman + docker-compose? For example, would there still be `_1` suffixes on each container name? I guess it doesn't really matter, as it's premature generalization to try to cover setups that we aren't going to use.

But about the specific case of our CoreOS server now, we are using systemd (in non-root user mode) to manage the starting and stopping of the AppBuilder stack there. The main reason was so that sites could be started automatically after a server reboot. And a nice side benefit being that we can use journalctl to view the combined console log output.

All that to say we should bear in mind that our use of podman on the CoreOS server may not necessarily generalize well to other cases.

File: README.md [hunk 1]
```diff
@@ -7,8 +7,104 @@ The server side runtime for our AppBuilder project.
...
+   ```bash
+   $ rm ./dbinit-compose.yml
+   ```
+1. Run db migrations
+   ```bash
+   $ ./migrate.sh
+   ```
+
+## Podman
+The runtime can be alternatively used with podman. Most scripts of the scripts already support
```
Comment (Hiro-Nakamura) [inline README.md:108]: "Most scripts of the scripts" ... didn't read smoothly.  Should that be "Most of the scripts ..."  ?

File: UP.sh [hunk 2]
```diff
@@ -24,17 +24,27 @@ then
...
 then
-    TestOveride="-c ./test/setup/ci-test.overide.yml"
+   if [[ -n $Test ]]
+   then
+     TestOverride="-f ./test/setup/ci-test.overide.yml"
+   fi
+   podman compose -f $File -f docker-compose.override.yml $TestOverride -p $STACKNAME up -d
+	# systemctl --user start pod-${STACKNAME}.service
+else
+   nohup node ab_system_monitor.js &> /dev/null &
```
Comment (Hiro-Nakamura) [inline UP.sh:41]: do we not run `ab_system_monitor.js` in a `podman` case?

File: dockerImageUpdate.js [hunk 3]
```diff
@@ -4,37 +4,53 @@
...
 const { exec } = require("child_process");
 const path = require("path");
 const fs = require("fs").promises;
+const migrationContainer = "update_script_db_migrations"; // ref to clean up the migration container
 
 // const cwd = process.cwd();
 
-const stack = ` ${process.env.STACKNAME}`;
-const stackNoSpace = stack.replace(" ", "");
+const stack = process.env.STACKNAME;
```
Comment (Hiro-Nakamura) [inline dockerImageUpdate.js:19]: I'm not very familiar with this, but are we no longer worried about someone putting " " in the STACKNAME?

File: examples/docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+      - SITE_URL
+      - OKTA_ENABLED
+      - OKTA_DOMAIN
+      - OKTA_CLIENT_ID
+      - OKTA_CLIENT_SECRET
+      # - FILE_PROCESSOR_PATH
+      # - FILE_PROCESSOR_UPLOAD_DIR
+      # - FILE_PROCESSOR_MAXBYTES
+      - RELAY_ENABLED
+      - RELAY_ENABLE
```
Comment (Hiro-Nakamura) [inline examples/docker-compose.yml:87]: Do we actually need both of these `RELAY_*` entries?  Seems like we could simplify this to one of the other.  Or if they represent different settings ... we should rename these to reflect the purpose better.

File: shell.sh [hunk 5]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+# Import ENV variables
+set -o allexport
+source .env
+set +o allexport
+
+ID=`${PLATFORM} container ls -f name=${STACKNAME}_api_sails --format={{.ID}}`
```
Comment (Hiro-Nakamura) [inline shell.sh:8]: Is this only allowing us to go into `api_sails`?  How about if I want to get to the cli of another service?

**Review by Hiro-Nakamura — CHANGES_REQUESTED** [review]
Just a few questions to think through.

File: UP.sh (hunk 2 above)
Comment (nh758) [reply UP.sh:41]: My preference would be to not to run it at all. It's meant to be used with the `bot_manager` service which is not really a part of our plans. We're not using it anywhere that I know of, and not sure it works the way we would want it to. I definitely don't want to be using bot manager on the Core-Os, without a use case and confidence that it works correctly.
Also don't see podman being used in many other places, so for now I don't think we need to run it.

File: dockerImageUpdate.js (hunk 3 above)
Comment (nh758) [reply dockerImageUpdate.js:19]: We don't do any checks elsewhere for spaces in the `STACKNAME`, so I 'd be really surprised if someone got here with that being the case. From what I can tell the, we first used the local `stack` variable with a space proceeding the stackname, my guess is this made composing commands easier at the time. At some point we added `stackNoSpace` which doesn't have the space preceding it, again my guess is this was for convienence composing commands in a slightly different way. Anyways at this point we can just the `STACKNAME` without adding a space in all cases.

File: shell.sh (hunk 5 above)
Comment (nh758) [reply shell.sh:8]: This script was added to `ab-production-runtime` at some point. I didn't modify this (and don't plan to). For me the underlying commands are easy enough that I don't think scripting it is worthwhile. But, if someone else wants to improve this they can.

**Comment by nh758** [comment]
> But about the specific case of our CoreOS server now, we are using systemd (in non-root user mode) to manage the starting and stopping of the AppBuilder stack there. The main reason was so that sites could be started automatically after a server reboot. And a nice side benefit being that we can use journalctl to view the combined console log output.

Thanks @joshAppDev, this is an area I don't understand that well. What commands is `systemd` service running, it must use some composition tool do bring up our ab services?  Do we use `systemd` on other servers with docker swarm, or they don't restart automatically?

File: UP.sh (hunk 2 above)
Comment (Hiro-Nakamura) [reply UP.sh:41]: That's fine.

If we no longer want the features `bot_manager` was supposed to provide for us, we should discuss taking it off the plate of our installed resources.  MIght be a good discussion for another time.

**Comment by joshAppDev** [comment]
> What commands is systemd service running, it must use some composition tool do bring up our ab services?

Yep, it just runs podman-compose.
```
# Compose up
ExecStart=/var/home/appdevcore/.local/bin/podman-compose -f /var/home/appdevcore/www/accounting2/docker-compose.yml -f /var/home/appdevcore/www/accounting2/docker-compose.override.yml up

# Compose down, remove containers and volumes
ExecStop=/var/home/appdevcore/.local/bin/podman-compose -f /var/home/appdevcore/www/accounting2/docker-compose.yml -f /var/home/appdevcore/www/accounting2/docker-compose.override.yml down
```
The systemd service files are located in `~/.config/systemd/user`.


>Do we use systemd on other servers with docker swarm, or they don't restart automatically?

It's only needed with Podman, as that runs in non-root mode. Docker takes care of restarting its stacks on reboot. Though we do have one or two sites using docker-compose instead of Docker swarm (Countly and maybe Lighthouse, I can't remember exactly now), and the auto-reboot for those has not been as reliable.

**Pull request merged**